docs: Add page on decentralized AI inference (#3407)
dfinity/portal: docs/developer-docs/ai/inference.mdx

---
keywords: [intermediate, concept, AI, ai, deAI, deai]
---

import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";

> # Decentralized AI Inference

Reviewer: ```suggestion
# Decentralized AI inference
```

<MarkdownChipRow labels={["Intermediate", "Concept", "DeAI" ]} />

It's possible for canister smart contracts to run inference in a number of ways, depending on the decentralization and performance requirements.

> ## 1. Inference on-chain

Reviewer: ```suggestion
## Inference on-chain
```

Currently, ICP supports on-chain inference of small models using AI libraries such as [Sonos Tract](https://github.com/sonos/tract) that compile to WebAssembly.
Check out the [image classification example](/docs/current/developer-docs/ai/ai-on-chain) to learn how it works.

> The long-term [vision of DeAI on ICP](https://internetcomputer.org/roadmap#Decentralized%20AI-start) is to support on-chain GPU compute to enable both training and inference of larger models.

Reviewer: This sentence is repeated from the Overview. It is not necessary to repeat the same exact wording here.

### Examples

> - [GPT2](https://github.com/modclub-app/rust-connect-py-ai-to-ic/tree/main/internet_computer/examples/gpt2): An example of GPT2 running on-chain. `Rust`

Reviewer: ```suggestion
- [GPT2](https://github.com/modclub-app/rust-connect-py-ai-to-ic/tree/main/internet_computer/examples/gpt2): An example of GPT2 running on-chain using `Rust`.
```

- [ELNA AI](https://github.com/elna-ai): A fully on-chain AI agent platform and marketplace. Supports both on-chain and off-chain LLMs. [Try it here](https://dapp.elna.ai/).

> - [Tensorflow on ICP](https://github.com/carlosarturoceron/decentAI): An Azle example that uses a pre-trained model for making predictions. `TypeScript`

Reviewer: ```suggestion
- [Tensorflow on ICP](https://github.com/carlosarturoceron/decentAI): An Azle example that uses TypeScript and a pre-trained model for making predictions. 
```

> - [ICGPT](https://github.com/icppWorld/icgpt): A React frontend that uses a C/C++ backend running an LLM fully on-chain. [Try it here](https://icgpt.icpp.world/). `TypeScript` `C++`

Reviewer: ```suggestion
- [ICGPT](https://github.com/icppWorld/icgpt): A React frontend that uses a C/C++ backend running an LLM fully on-chain. [Try it here](https://icgpt.icpp.world/).
```

Author: These tags are intentional, as it makes it easier to search for examples in specific languages.

Reviewer: This isn't a format that we use across the docs to tag languages, and in most instances the tags feel redundant since the language is most often included in the description of the project. Also, I'm not sure that adding these has any meaningful impact on the search results or search functionality. 

Alternatively, we can add UI tags for these languages that does align with how we format things across the docs.

> - [ArcMind AI](https://github.com/arcmindai/arcmindai): An autonomous agent using Chain of Thoughts for reasoning and actions. [Try it here](https://arcmindai.app). `Rust`

Reviewer: ```suggestion
- [ArcMind AI](https://github.com/arcmindai/arcmindai): An autonomous agent written in Rust using chain of thoughts for reasoning and actions. [Try it here](https://arcmindai.app).
```

### On-chain inference frameworks

> - [Sonos Tract](https://github.com/sonos/tract): An open-source AI inference engine written in Rust that supports ONNX, TensorFlow, PyTorch models and compiles to WebAssembly. `Rust`

Reviewer: ```suggestion
- [Sonos Tract](https://github.com/sonos/tract): An open-source AI inference engine written in Rust that supports ONNX, TensorFlow, and PyTorch models, and compiles to WebAssembly. 
```

- [MotokoLearn](https://github.com/ildefons/motokolearn): A Motoko package that enables on-chain machine learning.
  [The image classification example](https://github.com/dfinity/examples/tree/master/rust/image-classification) explains how to integrate it into a canister to run on ICP.

>   `Motoko`

Reviewer: ```suggestion

```

- [Rust-Connect-Py-AI-to-IC](https://github.com/jeshli/rust-connect-py-ai-to-ic): Open-source tool for deploying and running Python AI models on-chain using Sonos Tract.

>   `Rust`

Reviewer: ```suggestion

```

> - [Burn](https://github.com/tracel-ai/burn): An open-source deep learning framework written in Rust that supports ONNX, PyTorch models and compiles to WebAssembly.

Reviewer: ```suggestion
- [Burn](https://github.com/tracel-ai/burn): An open-source deep learning framework written in Rust that supports ONNX, and PyTorch models, and compiles to WebAssembly.
```

  [The MNIST example](https://github.com/smallstepman/ic-mnist) explains how to integrate it into a canister to run on ICP. [Try it here](https://jsi2g-jyaaa-aaaam-abnia-cai.icp0.io/).

>   `Rust`

Reviewer: ```suggestion
```

- [Candle](https://github.com/huggingface/candle): a minimalist ML framework for Rust that compiles to WebAssembly.
  [An AI chatbot example](https://github.com/ldclabs/ic-panda/tree/main/src/ic_panda_ai) shows how to run a Qwen 0.5B model in a canister on ICP.

>   `Rust`

Reviewer: ```suggestion
```

> ----

Reviewer: ```suggestion

```

Author: It would be useful for me to understand why a change is being suggested :) In this case, I saw the divider as something that visually helped with the organization of the page.

> ## 2. Inference on-device

Reviewer: ```suggestion
## Inference on-device
```

An alternative to running the model on-chain would be for the user to download the model from a canister smart contract, and the inference then happens on the user's device.
If the user trusts their own device, then they can trust that the inference ran correctly.
A disadvantage here is that the model needs to be downloaded to the user's device with corresponding drawbacks of less confidentiality of the model and decreased user experience due to increased latency.
ICP supports this use case for practically all existing models because a smart contract on ICP can store models up to 400GiB.

### Examples

>   - [DeVinci](https://github.com/patnorris/DecentralizedAIonIC): An in-browser AI chatbot that uses an open-source LLM model served from ICP. [Try it here](https://x6occ-biaaa-aaaai-acqzq-cai.icp0.io/). `Motoko`

Reviewer: ```suggestion
  - [DeVinci](https://github.com/patnorris/DecentralizedAIonIC): An in-browser AI chatbot that uses an open-source LLM model served from ICP. [Try it here](https://x6occ-biaaa-aaaai-acqzq-cai.icp0.io/). 
```

> ----

Reviewer: ```suggestion
```

> ## 3. Inference with HTTP calls

Reviewer: ```suggestion
## Inference with HTTP calls
```

> Smart contracts running on ICP can make [HTTP requests](https://internetcomputer.org/docs/current/tutorials/developer-journey/level-3/3.2-https-outcalls/) to Web2 services including OpenAI and Claude.

Reviewer: ```suggestion
Smart contracts running on ICP can make [HTTP requests through HTTP outcalls](/docs/current/developer-docs/smart-contracts/advanced-features/https-outcalls/https-outcalls-overview) to Web2 services including OpenAI and Claude.
```

### Examples

> - [Juno + OpenAI](https://github.com/peterpeterparker/juno-openai): An example using Juno and OpenAI to generate images from prompts. [Try it here](https://pycrs-xiaaa-aaaal-ab6la-cai.icp0.io/). `TypeScript` `Rust`

Reviewer: ```suggestion
- [Juno + OpenAI](https://github.com/peterpeterparker/juno-openai): An example using Juno and OpenAI to generate images from prompts. [Try it here](https://pycrs-xiaaa-aaaal-ab6la-cai.icp0.io/).
```
